src/librustc/lint/mod.rs

```diff
@@ -669,3 +685,33 @@ impl<'a, 'tcx> intravisit::Visitor<'tcx> for LintLevelMapBuilder<'a, 'tcx> {
 pub fn provide(providers: &mut Providers) {
     providers.lint_levels = lint_levels;
 }
+
+/// Returns whether `span` originates in a foreign crate's external macro.
+///
+/// This is used to test whether a lint should be entirely aborted above.
+fn in_external_macro(sess: &Session, span: Span) -> bool {
```

Comment:
Can you make this public so we can reuse it from clippy?

Reply:
Certainly!